pyro/contrib/cevae/__init__.py

```diff
@@ -0,0 +1,396 @@
+        :param int batch_size: Batch size. Defaults to 100.
+        :param float learning_rate: Learning rate. Defaults to 1e-3.
+        :param float learning_rate_decay: Learning rate decay over all epochs;
+            the per-step decay rate will depend on batch size and number of epochs
+            such that the initial learning rate will be ``learning_rate`` and the final
+            learning rate will be ``learning_rate * learning_rate_decay``.
+            Defaults to 0.1.
+        :param float weight_decay: Weight decay. Defaults to 1e-4.
+        :return: list of epoch losses
+        """
+        assert x.dim() == 2 and x.size(-1) == self.feature_dim
+        assert t.shape == x.shape[:1]
+        assert y.shape == y.shape[:1]
+
+        dataset = TensorDataset(x, t, y)
+        dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=True)
+        num_steps = num_epochs * len(dataloader)
+        optim = ClippedAdam({"lr": learning_rate,
+                             "weight_decay": weight_decay,
+                             "lrd": learning_rate_decay ** (1 / num_steps)})
+        svi = SVI(self.model, self.guide, optim, TraceCausalEffect_ELBO())
+        losses = []
+        for epoch in range(num_epochs):
+            loss = 0
+            for x, t, y in dataloader:
+                loss += svi.step(x, t, y, size=len(dataset)) / len(dataset)
+            logger.info("epoch {: >3d} loss = {:0.6g}".format(epoch, loss / len(dataloader)))
+            losses.append(loss)
+        return losses
+
+    @torch.no_grad()
+    def ite(self, x, num_samples=None):
+        r"""
+        Computes Individual Treatment Effect for a batch of data ``x``.
+
+        .. math::
+
+            ITE(x) = \mathbb E\bigl[ \mathbf y \mid \mathbf X=x, do(\mathbf t=1) \bigr]
+                   - \mathbb E\bigl[ \mathbf y \mid \mathbf X=x, do(\mathbf t=0) \bigr]
+
+        This has complexity ``O(len(x) * num_samples ** 2``.
+
+        :param ~torch.Tensor x: A batch of data.
+        :param int num_samples: The number of monte carlo samples.
+            Defaults to ``self.num_samples`` which defaults to ``100``.
+        :return: A ``len(x)``-sized tensor of estimated effects.
+        :rtype: ~torch.Tensor
+        """
+        if num_samples is None:
+            num_samples = self.num_samples
+        if not torch._C._get_tracing_state():
+            assert x.dim() == 2 and x.size(-1) == self.feature_dim
+
+        with pyro.plate("num_particles", num_samples, dim=-2):
+            with poutine.trace() as tr, poutine.block(hide=["y", "t"]):
+                self.guide(x)
+            with poutine.do(data=dict(t=torch.zeros(()))):
+                y0 = poutine.replay(self.model, tr.trace)(x)
+            with poutine.do(data=dict(t=torch.ones(()))):
+                y1 = poutine.replay(self.model, tr.trace)(x)
+        return (y1 - y0).mean(0)
+
+    def jit_trace(self):
```

Comment:
@neerajprad does this look like a reasonable interface?

Reply:
This looks good. As discussed, how about naming this `to_script_module()` or `get_script_module()`?

Reply:
done.